### DISCUSSION TITLE
setting Subcatchment Parameters for calibration purposes

### DISCUSSION THREAD

**Discussion by bjamali** [pyswmm/pyswmm#378] [Q&A] [answered]
It seems that Subcatchment parameters that can be set by user are limited to width,  area,  fracImperv, slope  and curbLength (see class SubcParams(Enum) in toolkitapi.py). 

For calibration of  a SWMM model it is usually necessary to change parameters such as previous and impervious roughness, infiltration rates, depression storage, etc. Is this functionality already available?

**Comment by jennwuu** [comment]
Are you looking to change these parameters before simulation starts?

**Comment by bjamali** [comment]
Yes, thats correct.

**Comment by jennwuu** [comment]
I don't think this feature is available in the SWMM toolkitapi, which is what pyswmm use. I would make the request here: https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/issues

**Comment by yangdong963** [comment]
Hello, I am studying pyswmm recently, and I have a question to ask you. The parameter of changing the impervious area of the subcatchment cannot be modified. Is it my code problem? ‘’‘S1.percent_impervious = 10’

**Comment by yangdong963** [comment]
code:print(s1.percent_impervious)
    s1.percent_impervious = 10
    print(s1.percent_impervious)
output:
0.29
0.29

**Comment by katmratliff** [comment]
Hi @yangdong963, can you please post a larger portion of your code? Thanks!

**Comment by yangdong963** [comment]
from pyswmm import Simulation, Subcatchments
with Simulation('./C1C2.inp') as sim:
    s1 = Subcatchments(sim)['S1']
    s1.slope = 100
    s1.percent_impervious = 0
    s1.percent_impervious = 0.22
    print(sub.percent_impervious)




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "OpenWaterAnalytics/pyswmm"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年7月22日(星期三) 晚上8:41
收件人:&nbsp;"OpenWaterAnalytics/pyswmm"<pyswmm@noreply.github.com&gt;;
抄送:&nbsp;"杨先森"<841003445@qq.com&gt;;"Mention"<mention@noreply.github.com&gt;;
主题:&nbsp;Re: [OpenWaterAnalytics/pyswmm] setting Subcatchment Parameters for calibration purposes (#246)





 
Hi @yangdong963, can you please post a larger portion of your code? Thanks!

**Comment by yangdong963** [comment]
from pyswmm import Simulation, Subcatchments
with Simulation('./C1C2.inp') as sim:
    s1 = Subcatchments(sim)['S1']
    s1.slope = 100
    print(s1.percent_impervious)
    s1.percent_impervious = 0.22
    print(s1.percent_impervious)
Hello, when I tried to modify the percent_impervious parameter in subcatchments, I found that this parameter cannot be modified, and parameters such as slope can be modified. My python environment is python 3.7. Thank you. Looking forward to your reply.
------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "OpenWaterAnalytics/pyswmm"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年7月22日(星期三) 晚上8:41
收件人:&nbsp;"OpenWaterAnalytics/pyswmm"<pyswmm@noreply.github.com&gt;;
抄送:&nbsp;"杨先森"<841003445@qq.com&gt;;"Mention"<mention@noreply.github.com&gt;;
主题:&nbsp;Re: [OpenWaterAnalytics/pyswmm] setting Subcatchment Parameters for calibration purposes (#246)





 
Hi @yangdong963, can you please post a larger portion of your code? Thanks!

**Comment by katmratliff** [comment]
thanks @yangdong963. As @bemcdonnell pointed out [here](https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/issues/298#issuecomment-662435305), this feature is not yet available. From the comments in the code, it looks like this may be a more challenging setter to implement, as it will require recalculation of other parameters after setting. I believe @jennwuu is working on some of the subcatchment parameters, but I'm not sure about this one (?)

**Comment by yangdong963** [comment]
Thank you for answering my question. In the process of calibrating the model, the parameters in subcatchments such as N-Imperv, N-perv, Dstore-imperv and other parameters are also empirical parameters that can be adjusted. I don’t know if you will consider them in the next development process. Yet?
One last word, thank you




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "OpenWaterAnalytics/pyswmm"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年7月22日(星期三) 晚上9:05
收件人:&nbsp;"OpenWaterAnalytics/pyswmm"<pyswmm@noreply.github.com&gt;;
抄送:&nbsp;"杨先森"<841003445@qq.com&gt;;"Mention"<mention@noreply.github.com&gt;;
主题:&nbsp;Re: [OpenWaterAnalytics/pyswmm] setting Subcatchment Parameters for calibration purposes (#246)





 
thanks @yangdong963. As @bemcdonnell pointed out here, this feature is not yet available. From the comments in the code, it looks like this may be a more challenging setter to implement, as it will require recalculation of other parameters after setting. I believe @jennwuu is working on some of the subcatchment parameters, but I'm not sure about this one (?)

**Comment by bemcdonnell** [comment]
@yangdong963, would you be interested in contributing these features?

**Comment by yangdong963** [comment]
Thank you, I'm just new to programming and I still don't have the ability to contribute




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "OpenWaterAnalytics/pyswmm"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年7月22日(星期三) 晚上9:18
收件人:&nbsp;"OpenWaterAnalytics/pyswmm"<pyswmm@noreply.github.com&gt;;
抄送:&nbsp;"杨先森"<841003445@qq.com&gt;;"Mention"<mention@noreply.github.com&gt;;
主题:&nbsp;Re: [OpenWaterAnalytics/pyswmm] setting Subcatchment Parameters for calibration purposes (#246)





 
@yangdong963, would you be interested in contributing these features?

**Comment by jennwuu** [comment]
@yangdong963 yes, N-Imperv, N-perv, Dstore-imperv, and Dstore-perv are parameters I am planning to expand in subcatchment getters and setters. 

I  did not evaluate the complexity of adding impervious fraction yet. I may include this parameter as well.

**Comment by tamaddun** [comment]
@yangdong963 Are you aware of any update on this?

**Comment by jennwuu** [comment]
@tamaddun I am still working on this feature. It's more complex than I originally thought and that's why it is delayed.

**Comment by tamaddun** [comment]
@jennwuu Thank you. Rooting for you!

**Comment by BuczynskiRafal** [comment]
Hi
I was trying to fix the percent_impervious setter today.
I thoroughly debugged the code looking for where the bug is. 
Unfortunately, I am not able to fix the operation of this stetter. 
It seems to me that I do not fully understand the operation of the following function form _solver.py file. 
```python
def subcatch_set_parameter(*args, **kwargs): # real signature unknown
    """
    Set a property value for specified subcatchment.
    
    Parameters
    ----------
    index: int
        The index of a subcatchment
    param: SM_SubcProperty
        The property type code (See :ref: SM_SubcProperty)
    value: double
        The new value of the subcatchment's property
    """
    pass
```

What exactly does it do with the passed arguments to function subcatch_set_parameter from solver.py file?

```python
def subcatch_set_parameter(index: "int", param: "SM_SubcProperty", value: "double") -> "int":
    r"""
    Set a property value for specified subcatchment.

    Parameters
    ----------
    index: int
        The index of a subcatchment
    param: SM_SubcProperty
        The property type code (See :ref: SM_SubcProperty)
    value: double
        The new value of the subcatchment's property

    """
    return _solver.subcatch_set_parameter(index, param, value)
```

I also noticed that changing the numbering in the SubcParams class as below setter works. This class in toolkitapi.py.  The simulation executes correctly. Can someone direct me why there is a problem with key 2 and where can I look for a solution? 

```python
class SubcParams(Enum):
    width = 0  # double
    area = 2  # double
    fracImperv = 1  # double
    slope = 3  # double
    curbLength = 4  # double
    # initBuildup = 5  # double
```

**Comment by BuczynskiRafal** [comment]
Okay, I see what the problem is. 
The validation in question is planned to be on the PySWMM side? I might be able to help implement this? 

```c
                break; //Subcatch[index].fracImperv; break;
                // Cannot Open Function just yet.  Need
                // to adjust some internal functions to
                // ensure parameters are recalculated
                // = MIN(value, 100.0) / 100.0; break;
```

**Comment by bemcdonnell** [comment]
@BuczynskiRafal, if you scroll down to the bottom of this page it will show you the links to the various repositories and how the project is architected. 

https://www.pyswmm.org/ 

I’m going to convert the issue to a discussion until we find units of work to work on

**Comment by suppi1610** [comment]
Hi everyone,
I am also stuck with the same problem, when I went through this discussion. I hope I'm not making any mistake in entering the code. And I just wanted to know if the function of percent_impervious is still functional or not?
Also if it's not functional yet, I guess it is better to remove it, from the pyswmm reference page (https://pyswmm.github.io/pyswmm/reference/api/api/pyswmm.subcatchments.Subcatchment.percent_impervious.html), temporarily until it's functional as it is misleading and wastes a lot of time.
Thank you.

**Comment by bemcdonnell** [comment, marked as answer]
@suppi1610, I think taking that out of the documentation is a good idea.  We have developed a solid work-around for this.

this uses our new `SimulationPreConfig` feature.  

```
[SUBCATCHMENTS]
;;                                         Total   Pcnt. Pcnt.  Curb
;;Name   Raingage         Outlet   Area    Imperv  Width Slope  Length
;;------ ---------------- -------- ------- ------- ----- ------ --------
S1       SCS_24h_Type_I_1in J1     1       100     500   0.5    0
```

```python
from pyswmm import Simulation, SimulationPreConfig, Subcatchments

# Create Config Handle
sim_conf = SimulationPreConfig()

# Specifying the update parameters
# Parameter Order:
# Section, Object ID, Parameter Index, New Value, Obj Row Num (optional)
sim_conf.add_update_by_token("SUBCATCHMENTS", "S1", 3, 47.231234)

with Simulation(<path-to-inp>, sim_preconfig = sim_conf) as sim:
    S1 = Subcatchments(sim)["S1"]
    print(S1.percent_impervious)

    for step in sim:
        pass
```

Have a look at the docs:

http://pyswmm.github.io/pyswmm/reference/api/pyswmm.simulation.SimulationPreConfig.html
